luatest/server.lua

```diff
@@ -547,24 +538,49 @@ function Server:exec(fn, args, options)
         error('net_box is not connected', 2)
     end
 
-    local ups = utils.upvalues(fn)
-    if next(ups) ~= nil then
+    local autorequired_pkgs = {'luatest'}
+    local passthrough_ups = {}
+    local other_ups = {}
+    for i = 1, debug.getinfo(fn, 'u').nups do
+        local name, value = debug.getupvalue(fn, i)
+        for _, pkg_name in ipairs(autorequired_pkgs) do
+            if value == package.loaded[pkg_name] then
+                passthrough_ups[name] = pkg_name
+                break
+            end
+        end
+        if not passthrough_ups[name] then
+            table.insert(other_ups, name)
+        end
+    end
+
+    if next(other_ups) ~= nil then
         local err = string.format(
             'bad argument #2 to exec (excess upvalues: %s)',
-            table.concat(ups, ', ')
+            table.concat(other_ups, ', ')
         )
         error(err, 2)
     end
 
-    return exec_tail(self:eval([[
-        local dump, args = ...
+    return exec_tail(self.net_box:eval([[
+        local dump, args, passthrough_ups = ...
         local fn = loadstring(dump)
+        local nups = debug.getinfo(fn, 'u').nups
+        for up_name, pkg_name in pairs(passthrough_ups) do
+            for i = 1, nups do
+                local name, value = debug.getupvalue(fn, i)
+                if name == up_name then
+                    debug.setupvalue(fn, i, require(pkg_name))
+                    break
+                end
+            end
+        end
```

Comment:
If I don't miss anything, we can look over upvalues and lookup the `passthrough_ups` table by the upvalue name (without the nested loop).

Reply:
Fixed.